test/phoenix/socket/transport_test.exs

```diff
@@ -288,5 +288,70 @@ defmodule Phoenix.Socket.TransportTest do
                |> fetch_query_params()
                |> Transport.connect_info(Endpoint, connect_info)
     end
+
+
+    defmodule EndpointWithCustomCSRFToken do
+      use Phoenix.Endpoint, otp_app: :phoenix
+
+      @session_config [
+        store: :cookie,
+        key: "_hello_key",
+        signing_salt: "change_me"
+      ]
+
+      def session_config, do: @session_config
+
+      plug Plug.Session, @session_config
+      plug :fetch_session
+      plug Plug.CSRFProtection, session_key: "_custom_csrf_token"
+      plug :put_session
+
+      defp put_session(conn, _) do
+        conn
+        |> put_session(:from_session, "123")
+        |> send_resp(200, Plug.CSRFProtection.get_csrf_token())
+      end
+    end
+
+    Application.put_env :phoenix, __MODULE__.EndpointWithCustomCSRFToken,
+      force_ssl: [],
+      url: [host: "host.com"],
+      check_origin: ["//endpoint.com"],
+      secret_key_base: @secret_key_base
+
+    test "loads the session with custom :csrf_token_session_key" do
+      {:ok, pid} = EndpointWithCustomCSRFToken.start_link()
```

Comment:
Instead of defining a new endpoint, you can try to emulate routes in the previous endpoint. Something like, set this particular session key based on `conn.path_info`.

Reply:
I can't figure out how to do this. The CSRF session token seems to be generated by a `before_send` callback defined on the `conn`. So I couldn't figure out how to make my own plug in the existing Endpoint that removes `"_csrf_token"` and adds my own based on path info.

Reply:
Hi @cjbottaro! I don't want to leave you hanging but I don't have answers for your questions right now and I would have to play with the code, but lacking time to do so at the moment, sorry.

Reply:
No worries at all. I'm just using my own branch for now, so no rush... 👍

Reply:
What about something like this:

```elixir
  defmodule Endpoint do
    use Phoenix.Endpoint, otp_app: :phoenix
    @session_config [
      store: :cookie,
      key: "_hello_key",
      signing_salt: "change_me"
    ]
    def session_config, do: @session_config
    plug Plug.Session, @session_config
    plug :fetch_session
    plug :put_csrf
    plug :put_session

    defp put_csrf(conn, _) do
      Plug.CSRFProtection.call(conn, session_key: Enum.join(conn.path_info, "_"))
    end

    defp put_session(conn, _) do
      conn
      |> put_session(:from_session, "123")
      |> send_resp(200, Plug.CSRFProtection.get_csrf_token())
    end
  end

```

Reply:
@Schultzer Cool, thanks, I got that working with a little fiddling. 👍

Do you have any opinions on the implementation of this feature? It works, but I feel it is hacky. The issue is that Phoenix doesn't store the _original_ session store options anywhere. It only stores the options after they have passed through `Plug.Session.Store.init/1`.

```elixir
orig_options = [csrf_token_key: "custom_key", ...]
init_options = MySessionStore.init(orig_options)
# ^ is what Phoenix stores, which isn't required to have the orig_options at all.
```

My hack does something like this...
```elixir
orig_options = [csrf_token_key: "custom_key", ...]
init_options = MySessionStore.init(orig_options)
{orig_options, init_options}
# ^ Phoenix stores a tuple of orig_options and init_options,
# that way we have access to csrf_token_key downstream.

cc @josevalim

Reply:
Actually, now that I articulated that though and coded it... it seems like a good solution, and not as hacky as I did it before. Returning both session_config and the results of init is more flexible in the long run, and it's conceptually simple.